src/k8s/pkg/k8sd/api/certs_refresh.go

```diff
@@ -0,0 +1,346 @@
+package api
+
+import (
+	"crypto/rand"
+	"crypto/x509/pkix"
+	"encoding/binary"
+	"fmt"
+	"net"
+	"net/http"
+	"os"
+	"path/filepath"
+
+	apiv1 "github.com/canonical/k8s/api/v1"
+	"github.com/canonical/k8s/pkg/k8sd/pki"
+	"github.com/canonical/k8s/pkg/k8sd/setup"
+	"github.com/canonical/k8s/pkg/log"
+	"github.com/canonical/k8s/pkg/snap"
+	snaputil "github.com/canonical/k8s/pkg/snap/util"
+	"github.com/canonical/k8s/pkg/utils"
+	pkiutil "github.com/canonical/k8s/pkg/utils/pki"
+	"github.com/canonical/lxd/lxd/response"
+	"github.com/canonical/microcluster/state"
+	v1 "k8s.io/api/certificates/v1"
+	corev1 "k8s.io/api/core/v1"
+)
+
+func (e *Endpoints) postRefreshCertsPlan(s *state.State, r *http.Request) response.Response {
+	snap := e.provider.Snap()
+	isWorker, err := snaputil.IsWorker(snap)
+	if err != nil {
+		return response.InternalError(fmt.Errorf("failed to check if node is a worker: %w", err))
+	}
+	if isWorker {
+		return refreshCertsPlanWorker(s, r, snap)
+	} else {
+		// TODO: Control Plane refresh
+		return response.InternalError(fmt.Errorf("not implemented yet"))
+	}
+
+}
+
+// refreshCertsPlanWorker generates the CSRs for the worker node and returns the seed and the names of the CSRs.
+func refreshCertsPlanWorker(s *state.State, r *http.Request, snap snap.Snap) response.Response {
+	log := log.FromContext(r.Context())
+
+	client, err := snap.KubernetesNodeClient("")
+	if err != nil {
+		return response.InternalError(err)
+	}
+
+	var seed int32
+
+	err = binary.Read(rand.Reader, binary.BigEndian, &seed)
+	if err != nil {
+		return response.InternalError(err)
+	}
+	seed = seed & 0x7FFFFFFF
+
+	log.Info("Generating CSRs for worker node")
+	log.Info("Generating Kubelet Serving Certificate Signing Request")
+
+	csrKubeletServing, pKeyKubeletServing, err := pkiutil.GenerateCSR(
+		pkix.Name{
+			CommonName:   fmt.Sprintf("system:node:%s", snap.Hostname()),
+			Organization: []string{"system:nodes"},
+		},
+		2048,
+		nil,
+		[]string{snap.Hostname()},
+		[]net.IP{net.ParseIP(s.Address().Hostname())},
+	)
+
+	_, err = client.CreateCertificateSigningRequest(
+		r.Context(),
+		fmt.Sprintf("k8sd-%d-worker-kubelet-serving", seed),
+		[]byte(csrKubeletServing),
+		[]v1.KeyUsage{v1.UsageDigitalSignature, v1.UsageKeyEncipherment, v1.UsageServerAuth},
+		[]string{"system:nodes"},
+		"k8sd.io/kubelet-serving",
+	)
+	if err != nil {
+		return response.InternalError(err)
+	}
```

Comment:
helpers are supposed to not just wrap the underlying calls but "group common operations". this whole process is such a thing, `client.CreateCertificateSigningRequest` on its own is not useful.